[D1] 🚀 Feature Request: Intercept the dev server error "binding DB of type d1 must have a valid `id` specified", make it clearer

### Describe the solution

If you attempt to run `wrangler dev --remote` with `preview_database_id` set to a random string that doesn't exist, you get a big scary error:

```
✘ [ERROR] Error on remote worker: ParseError: A request to the Cloudflare API (/accounts/xxxxxx/workers/scripts/d1-test/edge-preview) failed.

      at throwFetchError
  (/d1-test/node_modules/wrangler/wrangler-dist/cli.js:119636:17)
      at fetchResult
  (/d1-test/node_modules/wrangler/wrangler-dist/cli.js:119555:5)
      at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
      at async createPreviewToken
  (/d1-test/node_modules/wrangler/wrangler-dist/cli.js:148372:29)
      at async createWorkerPreview
  (/d1-test/node_modules/wrangler/wrangler-dist/cli.js:148393:17)
      at async start
  (/d1-test/node_modules/wrangler/wrangler-dist/cli.js:148962:34) {
    text: 'A request to the Cloudflare API
  (/accounts/xxxxxx/workers/scripts/d1-test/edge-preview) failed.',
    notes: [
      {
        text: 'binding DB of type d1 must have a valid `id` specified [code: 10021]'
      }
    ],
    location: undefined,
    kind: 'error',
    code: 10021
  }
```

The only way you'd get this error is if the database doesn't exist, so we should just tell folks to use a real database here.

Attempting to deploy a Worker with a fake ID throws the same error, so we could make that nicer too.

--- a/.changeset/fair-shoes-melt.md
+++ b/.changeset/fair-shoes-melt.md
@@ -2,4 +2,4 @@
 "wrangler": patch
 ---
 
-add user friendly error messages for d1 validation errors
+fix: ensure d1 validation errors render user friendly messages

--- a/.changeset/fair-shoes-melt.md
+++ b/.changeset/fair-shoes-melt.md
@@ -1,0 +1,5 @@
+---
+"wrangler": patch
+---
+
+add user friendly error messages for d1 validation errors

--- a/packages/wrangler/src/deploy/deploy.ts
+++ b/packages/wrangler/src/deploy/deploy.ts
@@ -719,8 +719,13 @@ See https://developers.cloudflare.com/workers/platform/compatibility-dates for m
 				) {
 					err.preventReport();
 
-					if(err.notes[0].text === "binding DB of type d1 must have a valid `id` specified [code: 10021]") {
-						throw new UserError("You must use a real database in the database_id configuration. You can find your databases using 'wrangler d1 list', or read how to develop locally with D1 here: https://developers.cloudflare.com/d1/configuration/local-development");
+					if (
+						err.notes[0].text ===
+						"binding DB of type d1 must have a valid `id` specified [code: 10021]"
+					) {
+						throw new UserError(
+							"You must use a real database in the database_id configuration. You can find your databases using 'wrangler d1 list', or read how to develop locally with D1 here: https://developers.cloudflare.com/d1/configuration/local-development"
+						);
 					}
 
 					const maybeNameToFilePath = (moduleName: string) => {

--- a/packages/wrangler/src/deploy/deploy.ts
+++ b/packages/wrangler/src/deploy/deploy.ts
@@ -719,6 +719,10 @@ See https://developers.cloudflare.com/workers/platform/compatibility-dates for m
 				) {
 					err.preventReport();
 
+					if(err.notes[0].text === "binding DB of type d1 must have a valid `id` specified [code: 10021]") {
+						throw new UserError("You must use a real database in the database_id configuration. You can find your databases using 'wrangler d1 list', or read how to develop locally with D1 here: https://developers.cloudflare.com/d1/configuration/local-development");
+					}
+
 					const maybeNameToFilePath = (moduleName: string) => {
 						// If this is a service worker, always return the entrypoint path.
 						// Service workers can't have additional JavaScript modules.

--- a/packages/wrangler/src/dev/remote.tsx
+++ b/packages/wrangler/src/dev/remote.tsx
@@ -29,6 +29,7 @@ import type {
 	CfWorkerContext,
 	CfWorkerInit,
 } from "../deployment-bundle/worker";
+import type { ParseError } from "../parse";
 import type { AssetPaths } from "../sites";
 import type { ChooseAccountItem } from "../user";
 import type {
@@ -37,7 +38,6 @@ import type {
 	CfPreviewToken,
 } from "./create-worker-preview";
 import type { EsbuildBundle } from "./use-esbuild";
-import type { ParseError } from "../parse";
 
 interface RemoteProps {
 	name: string | undefined;
@@ -341,15 +341,14 @@ export function useWorker(
 			// give them friendly instructions
 			if ((err as unknown as { code: string }).code !== "ABORT_ERR") {
 				// code 10049 happens when the preview token expires
-				if(err.code === 10049) {
+				if (err.code === 10049) {
 					logger.log("Preview token expired, fetching a new one");
 
 					// since we want a new preview token when this happens,
 					// lets increment the counter, and trigger a rerun of
 					// the useEffect above
 					setRestartCounter((prevCount) => prevCount + 1);
-				}
-				else if(!handleUserFriendlyError(err, props.accountId)) {
+				} else if (!handleUserFriendlyError(err, props.accountId)) {
 					logger.error("Error on remote worker:", err);
 				}
 			}
@@ -524,10 +523,9 @@ export async function getRemotePreviewToken(props: RemoteProps) {
 		// give them friendly instructions
 		if ((err as unknown as { code: string })?.code !== "ABORT_ERR") {
 			// code 10049 happens when the preview token expires
-			if(err.code === 10049) {
+			if (err.code === 10049) {
 				logger.log("Preview token expired, restart server to fetch a new one");
-			}
-			else if(!handleUserFriendlyError(err, props.accountId)) {
+			} else if (!handleUserFriendlyError(err, props.accountId)) {
 				helpIfErrorIsSizeOrScriptStartup(err, props.bundle?.dependencies || {});
 				logger.error("Error on remote worker:", err);
 			}
@@ -679,14 +677,17 @@ function ChooseAccount(props: {
  * @returns if the error was handled or not
  */
 function handleUserFriendlyError(error: ParseError, accountId?: string) {
-	switch((error as unknown as { code: number }).code) {
+	switch ((error as unknown as { code: number }).code) {
 		// code 10021 is a validation error
 		case 10021: {
 			// if it is the following message, give a more user friendly
 			// error, otherwise do not handle this error in this function
-			if(error.notes[0].text === "binding DB of type d1 must have a valid `id` specified [code: 10021]") {
+			if (
+				error.notes[0].text ===
+				"binding DB of type d1 must have a valid `id` specified [code: 10021]"
+			) {
 				const errorMessage =
-				"Error: You must use a real database in the preview_database_id configuration.";
+					"Error: You must use a real database in the preview_database_id configuration.";
 				const solutionMessage =
 					"You can find your databases using 'wrangler d1 list', or read how to develop locally with D1 here:";
 				const documentationLink = `https://developers.cloudflare.com/d1/configuration/local-development`;
@@ -707,11 +708,11 @@ function handleUserFriendlyError(error: ParseError, accountId?: string) {
 				"Error: You need to register a workers.dev subdomain before running the dev command in remote mode";
 			const solutionMessage =
 				"You can either enable local mode by pressing l, or register a workers.dev subdomain here:";
-			const onboardingLink = (accountId)?(`https://dash.cloudflare.com/${accountId}/workers/onboarding`):("https://dash.cloudflare.com/?to=/:account/workers/onboarding");
+			const onboardingLink = accountId
+				? `https://dash.cloudflare.com/${accountId}/workers/onboarding`
+				: "https://dash.cloudflare.com/?to=/:account/workers/onboarding";
 
-			logger.error(
-				`${errorMessage}\n${solutionMessage}\n${onboardingLink}`
-			);
+			logger.error(`${errorMessage}\n${solutionMessage}\n${onboardingLink}`);
 
 			return true;
 		}

--- a/packages/wrangler/src/dev/remote.tsx
+++ b/packages/wrangler/src/dev/remote.tsx
@@ -684,20 +684,25 @@ function handleUserFriendlyError(error: FetchError, accountId?: string) {
 	}
 
 	switch(error.code) {
-		// code 10021 happens when the provided preview database
-		// id does not actually exist
+		// code 10021 is a validation error
 		case 10021: {
-			const errorMessage =
-			"Error: You must use a real database in the preview_database_id configuration.";
-			const solutionMessage =
-				"You can find your databases using 'wrangler d1 list', or read how to develop locally with D1 here:";
-			const documentationLink = `https://developers.cloudflare.com/d1/configuration/local-development`;
+			// if it is the following message, give a more user friendly
+			// error, otherwise do not handle this error in this function
+			if(error.message === "binding DB of type d1 must have a valid `id` specified") {
+				const errorMessage =
+				"Error: You must use a real database in the preview_database_id configuration.";
+				const solutionMessage =
+					"You can find your databases using 'wrangler d1 list', or read how to develop locally with D1 here:";
+				const documentationLink = `https://developers.cloudflare.com/d1/configuration/local-development`;
 
-			logger.error(
-				`${errorMessage}\n${solutionMessage}\n${documentationLink}`
-			);
+				logger.error(
+					`${errorMessage}\n${solutionMessage}\n${documentationLink}`
+				);
 
-			return true;
+				return true;
+			}
+
+			return false;
 		}
 
 		// for error 10063 (workers.dev subdomain required)

--- a/packages/wrangler/src/dev/remote.tsx
+++ b/packages/wrangler/src/dev/remote.tsx
@@ -37,6 +37,7 @@ import type {
 	CfPreviewToken,
 } from "./create-worker-preview";
 import type { EsbuildBundle } from "./use-esbuild";
+import { FetchError } from "../cfetch";
 
 interface RemoteProps {
 	name: string | undefined;
@@ -336,29 +337,20 @@ export function useWorker(
 		start().catch((err) => {
 			// we want to log the error, but not end the process
 			// since it could recover after the developer fixes whatever's wrong
-			if ((err as { code: string }).code !== "ABORT_ERR") {
-				// instead of logging the raw API error to the user,
-				// give them friendly instructions
-				// for error 10063 (workers.dev subdomain required)
-				if (err.code === 10063) {
-					const errorMessage =
-						"Error: You need to register a workers.dev subdomain before running the dev command in remote mode";
-					const solutionMessage =
-						"You can either enable local mode by pressing l, or register a workers.dev subdomain here:";
-					const onboardingLink = `https://dash.cloudflare.com/${props.accountId}/workers/onboarding`;
-					logger.error(
-						`${errorMessage}\n${solutionMessage}\n${onboardingLink}`
-					);
-				} else if (err.code === 10049) {
-					logger.log("Preview token expired, fetching a new one");
-					// code 10049 happens when the preview token expires
-					// since we want a new preview token when this happens,
-					// lets increment the counter, and trigger a rerun of
-					// the useEffect above
-					setRestartCounter((prevCount) => prevCount + 1);
-				} else {
-					logger.error("Error on remote worker:", err);
-				}
+			// instead of logging the raw API error to the user,
+			// give them friendly instructions
+
+			// code 10049 happens when the preview token expires
+			if(err.code === 10049) {
+				logger.log("Preview token expired, fetching a new one");
+
+				// since we want a new preview token when this happens,
+				// lets increment the counter, and trigger a rerun of
+				// the useEffect above
+				setRestartCounter((prevCount) => prevCount + 1);
+			}
+			else if(!handleUserFriendlyError(err, props.accountId)) {
+				logger.error("Error on remote worker:", err);
 			}
 		});
 
@@ -525,24 +517,18 @@ export async function getRemotePreviewToken(props: RemoteProps) {
 		return workerPreviewToken;
 	}
 	return start().catch((err) => {
-		if ((err as { code?: string })?.code !== "ABORT_ERR") {
-			// instead of logging the raw API error to the user,
-			// give them friendly instructions
-			// for error 10063 (workers.dev subdomain required)
-			if (err?.code === 10063) {
-				const errorMessage =
-					"Error: You need to register a workers.dev subdomain before running the dev command in remote mode";
-				const solutionMessage =
-					"You can either enable local mode by pressing l, or register a workers.dev subdomain here:";
-				const onboardingLink = `https://dash.cloudflare.com/${props.accountId}/workers/onboarding`;
-				logger.error(`${errorMessage}\n${solutionMessage}\n${onboardingLink}`);
-			} else if (err?.code === 10049) {
-				// code 10049 happens when the preview token expires
-				logger.log("Preview token expired, restart server to fetch a new one");
-			} else {
-				helpIfErrorIsSizeOrScriptStartup(err, props.bundle?.dependencies || {});
-				logger.error("Error on remote worker:", err);
-			}
+		// we want to log the error, but not end the process
+		// since it could recover after the developer fixes whatever's wrong
+		// instead of logging the raw API error to the user,
+		// give them friendly instructions
+
+		// code 10049 happens when the preview token expires
+		if(err.code === 10049) {
+			logger.log("Preview token expired, restart server to fetch a new one");
+		}
+		else if(!handleUserFriendlyError(err, props.accountId)) {
+			helpIfErrorIsSizeOrScriptStartup(err, props.bundle?.dependencies || {});
+			logger.error("Error on remote worker:", err);
 		}
 	});
 }
@@ -683,4 +669,54 @@ function ChooseAccount(props: {
 			/>
 		</>
 	);
+}
+
+/**
+ * A switch for handling thrown error mappings to user friendly
+ * messages, does not perform any logic other than logging errors.
+ * @returns if the error was handled or not
+ */
+function handleUserFriendlyError(error: FetchError, accountId?: string) {
+	// simulate aborts as handled errors by this function, but don't
+	// log anything
+	if ((error.code as unknown as string) === "ABORT_ERR") {
+		return true;
+	}
+
+	switch(error.code) {
+		// code 10021 happens when the provided preview database
+		// id does not actually exist
+		case 10021: {
+			const errorMessage =
+			"Error: You must use a real database in the preview_database_id configuration.";
+			const solutionMessage =
+				"You can find your databases using 'wrangler d1 list', or read how to develop locally with D1 here:";
+			const documentationLink = `https://developers.cloudflare.com/d1/configuration/local-development`;
+
+			logger.error(
+				`${errorMessage}\n${solutionMessage}\n${documentationLink}`
+			);
+
+			return true;
+		}
+
+		// for error 10063 (workers.dev subdomain required)
+		case 10063: {
+			const errorMessage =
+				"Error: You need to register a workers.dev subdomain before running the dev command in remote mode";
+			const solutionMessage =
+				"You can either enable local mode by pressing l, or register a workers.dev subdomain here:";
+			const onboardingLink = (accountId)?(`https://dash.cloudflare.com/${accountId}/workers/onboarding`):("https://dash.cloudflare.com/?to=/:account/workers/onboarding");
+
+			logger.error(
+				`${errorMessage}\n${solutionMessage}\n${onboardingLink}`
+			);
+
+			return true;
+		}
+
+		default: {
+			return false;
+		}
+	}
 }

--- a/packages/wrangler/src/dev/remote.tsx
+++ b/packages/wrangler/src/dev/remote.tsx
@@ -37,7 +37,7 @@ import type {
 	CfPreviewToken,
 } from "./create-worker-preview";
 import type { EsbuildBundle } from "./use-esbuild";
-import { FetchError } from "../cfetch";
+import type { FetchError } from "../cfetch";
 
 interface RemoteProps {
 	name: string | undefined;

--- a/packages/wrangler/src/dev/remote.tsx
+++ b/packages/wrangler/src/dev/remote.tsx
@@ -37,7 +37,7 @@ import type {
 	CfPreviewToken,
 } from "./create-worker-preview";
 import type { EsbuildBundle } from "./use-esbuild";
-import type { FetchError } from "../cfetch";
+import type { ParseError } from "../parse";
 
 interface RemoteProps {
 	name: string | undefined;
@@ -676,19 +676,19 @@ function ChooseAccount(props: {
  * messages, does not perform any logic other than logging errors.
  * @returns if the error was handled or not
  */
-function handleUserFriendlyError(error: FetchError, accountId?: string) {
+function handleUserFriendlyError(error: ParseError, accountId?: string) {
 	// simulate aborts as handled errors by this function, but don't
 	// log anything
-	if ((error.code as unknown as string) === "ABORT_ERR") {
+	if ((error as unknown as { code: string }).code === "ABORT_ERR") {
 		return true;
 	}
 
-	switch(error.code) {
+	switch((error as unknown as { code: number }).code) {
 		// code 10021 is a validation error
 		case 10021: {
 			// if it is the following message, give a more user friendly
 			// error, otherwise do not handle this error in this function
-			if(error.message === "binding DB of type d1 must have a valid `id` specified") {
+			if(error.notes[0].text === "binding DB of type d1 must have a valid `id` specified [code: 10021]") {
 				const errorMessage =
 				"Error: You must use a real database in the preview_database_id configuration.";
 				const solutionMessage =

--- a/packages/wrangler/src/dev/remote.tsx
+++ b/packages/wrangler/src/dev/remote.tsx
@@ -339,18 +339,19 @@ export function useWorker(
 			// since it could recover after the developer fixes whatever's wrong
 			// instead of logging the raw API error to the user,
 			// give them friendly instructions
+			if ((err as unknown as { code: string }).code !== "ABORT_ERR") {
+				// code 10049 happens when the preview token expires
+				if(err.code === 10049) {
+					logger.log("Preview token expired, fetching a new one");
 
-			// code 10049 happens when the preview token expires
-			if(err.code === 10049) {
-				logger.log("Preview token expired, fetching a new one");
-
-				// since we want a new preview token when this happens,
-				// lets increment the counter, and trigger a rerun of
-				// the useEffect above
-				setRestartCounter((prevCount) => prevCount + 1);
-			}
-			else if(!handleUserFriendlyError(err, props.accountId)) {
-				logger.error("Error on remote worker:", err);
+					// since we want a new preview token when this happens,
+					// lets increment the counter, and trigger a rerun of
+					// the useEffect above
+					setRestartCounter((prevCount) => prevCount + 1);
+				}
+				else if(!handleUserFriendlyError(err, props.accountId)) {
+					logger.error("Error on remote worker:", err);
+				}
 			}
 		});
 
@@ -521,14 +522,15 @@ export async function getRemotePreviewToken(props: RemoteProps) {
 		// since it could recover after the developer fixes whatever's wrong
 		// instead of logging the raw API error to the user,
 		// give them friendly instructions
-
-		// code 10049 happens when the preview token expires
-		if(err.code === 10049) {
-			logger.log("Preview token expired, restart server to fetch a new one");
-		}
-		else if(!handleUserFriendlyError(err, props.accountId)) {
-			helpIfErrorIsSizeOrScriptStartup(err, props.bundle?.dependencies || {});
-			logger.error("Error on remote worker:", err);
+		if ((err as unknown as { code: string })?.code !== "ABORT_ERR") {
+			// code 10049 happens when the preview token expires
+			if(err.code === 10049) {
+				logger.log("Preview token expired, restart server to fetch a new one");
+			}
+			else if(!handleUserFriendlyError(err, props.accountId)) {
+				helpIfErrorIsSizeOrScriptStartup(err, props.bundle?.dependencies || {});
+				logger.error("Error on remote worker:", err);
+			}
 		}
 	});
 }
@@ -677,12 +679,6 @@ function ChooseAccount(props: {
  * @returns if the error was handled or not
  */
 function handleUserFriendlyError(error: ParseError, accountId?: string) {
-	// simulate aborts as handled errors by this function, but don't
-	// log anything
-	if ((error as unknown as { code: string }).code === "ABORT_ERR") {
-		return true;
-	}
-
 	switch((error as unknown as { code: number }).code) {
 		// code 10021 is a validation error
 		case 10021: {

